c++/include/orc/OrcFile.hh

```diff
@@ -58,10 +62,38 @@ namespace orc {
      */
     virtual void read(void* buf, uint64_t length, uint64_t offset) = 0;
 
+    /**
+     * Read data asynchronously into the buffer. The buffer is allocated by the caller.
+     * @param buf the buffer to read into
+     * @param length the number of bytes to read.
+     * @param offset the position in the stream to read from.
+     * @return a future that will be set to the number of bytes read when the read is complete.
+     */
+    virtual std::future<void> readAsync(void* buf, uint64_t length, uint64_t offset) {
+      return std::async(std::launch::async,
+                        [this, buf, length, offset] { this->read(buf, length, offset); });
+    }
+
     /**
      * Get the name of the stream for error messages.
      */
     virtual const std::string& getName() const = 0;
+
+   protected:
```

Comment:
`protected` is indeed visible to downstream users. BTW, it looks weird that we need to pass a memory pool to a read function. Therefore I think `readAsyncInternal` is ill-designed. Together with my comment to your comment of the benchmark, please remove this function and stick to `readAsync`.